docs/proposals/scheduling/20220629-schedule-device-in-card-level.md

```diff
@@ -0,0 +1,268 @@
+
+3. runtime-proxy use scheduler device schedule result to overwrite kubelet\device-plugin card assign result
+
+### User Stories
+
+#### Story NVLink
+NVLink
+
+![image](/docs/images/nvlink.jpg)
+
+through the picture we can find that although there's 8 card in A100\V100 node, the data-transport-speed between cards 
+are not same. when a pod need multi-gpu-cards, we have chance to select a better placement with largest data-transport-speed. 
+
+#### Story MIG
+![image](/docs/images/mig.jpg)
+
+A100 gpu card support mig function to split a card into different size of ci\gi, which provide isolate ability.
+in some scenarios, we need create ci\gi dynamically with pod-scheduling, for example, some pods need 1c.4g.40gb, some pods
+need 1c.2g.20gb, we can't expect the pods requirements in advance, so there is a demand for scheduler to make the 
+best placement for ci\gi-create and schedule pod in it.
+
+#### Story Oversold
+in some cases, we find although some cards free is zero in scheduler account, but we observed that the physical
+utilization of gpu-mem\gpu-core still has buffer, so we may schedule some low-priority jobs to use this buffer, and
+when buffer disappear, we should kill these jobs to avoid affecting normal jobs. because the utilization is different
+bewteen cards, so scheduler must schedule based card level.
+     
+
+## Implementation Details
+
+#### Scheduler
+target:
+1. we need a new data structure to describe each device card resources and healthy status on one node
+2. we need new plugins to execute filter\reserve\bind stage
+3. support user to apply any ratio of one card, like 50%/25%
+4. automatically recognize different kind devices. when a new device added, we don't need modify any code
+5. schedule-result will be written on pod annotations.
+
+detail:
+
+1.device resoruce dimension:
+For GPU, we have 3 dimensions:gpu-core-ratio\gpu-mem-ratio\gpu-mem, this design is for future node isolation feature. 
+
+   gpu-core-ratio means calculation percentage of one card.
+
+   gpu-mem-ratio means memory percentage of one card. 
+
+   gpu-mem is real memory-byte of pod usage. 
+
+   we assume one card's total resource is {gpu-core-ratio:100, gpu-mem-ratio:100, gpu-mem:xxGB}, xxGB is reported by device 
+reporter. For other devices like RDMA\FPGA, we temporary only has rdma\fpga demension, total resource is 
+{rdma:100}, {fpga:100}.
+
+2.user apply device resource:
+
+   a. if user apply {nvidia.com/gpu:1}, we translate to {gpu-core-ratio:100, gpu-mem-ratio:100， gpu-mem:0GB}. gpu-mem 0GB
+is because different gpu model has different gpu-memory. so when we filter different nodes, we will determine the 
+real gpu-mem value by node gpu memory * pod's gpu-mem-ratio in real time.
+
+   b. if user apply {koordinator/gpu:100}, we translate to {gpu-core-ratio:100, gpu-mem-ratio:100， gpu-mem:0GB}.
+
+   c. if user apply {koordinator/gpu-core:50, koordinator/gpu-mem-ratio:50}, we translate to {gpu-core-ratio:50, 
```

Comment:
Memory supports ratio or absolute value allocation, but (cuda) core supports ratio allocation only, which can be confusing. Keep a consistent description, so that it can be satisfied in the future when it needs to be scheduled according to the absolute cuda core number? user apply resource like this:
{koordinator/gpu-ratio:100}
{koordinator/gpu-core-ratio:20, koordinator/gpu-mem-ratio:50}

{koordinator/gpu-core:1024, koordinator/gpu-mem:5GB} // reserved for future consideration